New blog: Why we should write tests for our production code (#246)
incubyte/incubyte.github.io: content/blog/why-write-tests-for-production-code.md

+++

> title = "Why We Should Write Tests for Our Production Code"

Reviewer: Can we change the blog title a bit
Here are some suggestions from gpt

![image](https://github.com/user-attachments/assets/db266362-03d9-4b31-80c3-e89d474d9dd7)

Author: Yes sure we can change it, what about the last one `Testing Production Code: Why It's Essential`?

Reviewer 2: To make it a bit more engaging: "Its 2025 and you should not write code without tests", goes with the suggestion in my previous comment

slug = "why-we-should-write-tests-for-our-production-code"
date = 2025-01-03T11:10:20+05:30
image = "/images/2025/why-write-tests-for-production-code/header.svg"
draft = false
authors = ["Tilak Patidar"]
description = ""
tags = ["TDD", "Software Craftsmanship"]
categories = ["Playbook", "Software Craftsmanship"]
type = ""
+++

> _(Insights from Clean Code by Robert C. Martin, Chapter 9: Unit Tests)_

Reviewer 2: Lets link the book / chapter here

Software development practices have evolved significantly over the decades.
One of the most notable transformations has been in how developers approach testing.
The journey from manual, throw-away tests to systematic, automated testing practices reflects a
deeper understanding of what makes code maintainable and reliable.

> Back in the mid-90s, unit tests were often just temporary code snippets used to verify if programs "worked."

Reviewer 2: ```suggestion
Back in the mid-90s, unit tests were often just temporary code snippets used to verify if programs "worked"
```

tests every day. Do the math, those daily tests add up to hundreds in a month and thousands in a year. Soon
enough, the test code becomes almost as large as the production code itself.

But this isn't a bad thing. These tests become your safety net, your documentation, and your design tool all
rolled into one. When you write tests first, you naturally create code that's easier to test. The code becomes
more modular, with cleaner interfaces and fewer dependencies. It's like having a blueprint before building a
house, you catch potential problems before they become expensive mistakes.

The real beauty of TDD shows up when you need to make changes. Instead of wondering "what might break if I
change this?", you can modify code confidently. If something breaks, your tests will tell you immediately.
No more hours spent manually testing the entire application after every small change.

Yes, writing tests first might feel slower initially. But it's like stopping to sharpen your axe before
cutting down a tree, the small upfront investment saves much more time in the long run. You catch bugs

> earlier when they're cheaper to fix, you spend less time debugging, and your code stays clean and maintainable.

Reviewer 2: ```suggestion
earlier when they are cheaper to fix, you spend less time debugging, and your code stays clean and maintainable.
```

> Remember though, just having a lot of tests isn't enough. These tests need to be clean, readable, and

Reviewer 2: ```suggestion
Remember though, just having a lot of tests isn't enough. These tests need to be purposeful, readable, and
```

well-maintained. They're as much a part of your codebase as your production code, and they deserve the same
level of care and attention.

### Writing Clean Tests

What makes a test "clean"? It all comes down to one word: readability. A test should tell a story about the code.
When someone reads your test, they should instantly understand what's being tested and what's expected to
happen.

> Clean tests follow a simple pattern: build, operate, check. First, you set up your test data. Then you do

Reviewer 2: ```suggestion
Clean tests follow a simple pattern: build, operate, check AKA arrange, act, assert. First, you set up your test data. Then you do
```

+++

> title = "Its 2025 and you should not write code without tests"

Reviewer 2: Lets make the title: Title Case
